File: Chap.77.어려운문제들/14500.테트로미노/solve.fastest.py
def dfs(d, s, i, j, n, m, A):
    print(d, i, j, s)
    if d == 0:
        return s
    else:
        right = 0 if i == n - 1 else dfs(d - 1, s + A[i+1][j], i+1, j, n, m, A)
        top = 0 if i == 0 else dfs(d - 1, s + A[i-1][j], i-1, j, n, m, A)
        bottom = 0 if j == m - 1 else dfs(d - 1, s + A[i][j+1], i, j+1, n, m, A)
        return max(right, top, bottom)

def solve(n, m, A):
    opt = 0
    for i in range(n):
        for j in range(n):
            opt = max(opt, dfs(3, A[i][j], i, j, n, m, A))
    # ㅗ ㅜ
    h, w = 2, 3 
    for i in range(n - h + 1):
        for j in range(m - w + 1):
            opt = max(opt, A[i][j]+A[i][j+1]+A[i][j+2]+A[i+1][j+1]);
            opt = max(opt, A[i+1][j]+A[i+1][j+1]+A[i+1][j+2]+A[i][j+1]);
    # ㅏ ㅓ
    h, w = 3, 2
    for i in range(n - h + 1):
        for j in range(m - w + 1):
            opt = max(opt, A[i][j]+A[i+1][j]+A[i+2][j]+A[i+1][j+1]);
            opt = max(opt, A[i][j+1]+A[i+1][j+1]+A[i+2][j+1]+A[i+1][j]);
    return opt

import sys
input = sys.stdin.readline

n, m = map(int, input().split())
A = [list(map(int, input().split())) for _ in range(n)]
print(solve(n, m, A))

File: Chap.77.어려운문제들/14500.테트로미노/solve.fastest.c
#include <stdio.h>
#define max(x,y) (x>y)?x:y

int main() {
    int n, m, opt, i, j, h, w;
    int A[500][500];
    scanf("%d %d", &n, &m);
    for (i = 0; i < n; i++)
        for (j = 0; j < m; j++)
            scanf("%d", &A[i][j]);
            
    opt = 0;

    for (i = 0; i < n; i++)
        for (j = 0; j < m; j++) {

            h = 2; w = 2;
            if (i <= n - h && j <= m - w)
                opt = max(opt, A[i][j]+A[i+1][j]+A[i][j+1]+A[i+1][j+1]);

            h = 1; w = 4;
            if (i <= n - h && j <= m - w)
                opt = max(opt, A[i][j]+A[i][j+1]+A[i][j+2]+A[i][j+3]);

            h = 4; w = 1;
            if (i <= n - h && j <= m - w)
                opt = max(opt, A[i][j]+A[i+1][j]+A[i+2][j]+A[i+3][j]);

            h = 2; w = 3;
            if (i <= n - h && j <= m - w) {
                opt = max(opt, A[i][j]+A[i][j+1]+A[i+1][j+1]+A[i+1][j+2]);
                opt = max(opt, A[i+1][j]+A[i+1][j+1]+A[i][j+1]+A[i][j+2]);
                opt = max(opt, A[i][j]+A[i][j+1]+A[i][j+2]+A[i+1][j+1]);
                opt = max(opt, A[i+1][j]+A[i+1][j+1]+A[i+1][j+2]+A[i][j+1]);
                opt = max(opt, A[i][j]+A[i+1][j]+A[i+1][j+1]+A[i+1][j+2]);
                opt = max(opt, A[i][j]+A[i][j+1]+A[i][j+2]+A[i+1][j+2]);
                opt = max(opt, A[i][j]+A[i][j+1]+A[i][j+2]+A[i+1][j]);
                opt = max(opt, A[i+1][j]+A[i+1][j+1]+A[i+1][j+2]+A[i][j+2]);
            }

            h = 3; w = 2;
            if (i <= n - h && j <= m - w) {            
                opt = max(opt, A[i][j]+A[i+1][j]+A[i+1][j+1]+A[i+2][j+1]);
                opt = max(opt, A[i][j+1]+A[i+1][j]+A[i+1][j+1]+A[i+2][j]);
                opt = max(opt, A[i][j]+A[i+1][j]+A[i+2][j]+A[i+1][j+1]);
                opt = max(opt, A[i][j+1]+A[i+1][j+1]+A[i+2][j+1]+A[i+1][j]);
                opt = max(opt, A[i][j]+A[i+1][j]+A[i+2][j]+A[i+2][j+1]);
                opt = max(opt, A[i][j]+A[i][j+1]+A[i+1][j]+A[i+2][j]);
                opt = max(opt, A[i][j+1]+A[i+1][j+1]+A[i+2][j+1]+A[i+2][j]);
                opt = max(opt, A[i][j]+A[i][j+1]+A[i+1][j+1]+A[i+2][j+1]);
            }
        }

    printf("%d", opt);
}